pkg/compute/build.go

```diff
@@ -129,17 +124,168 @@ func (c *BuildCommand) Exec(in io.Reader, out io.Writer) (err error) {
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// Create implicit top-level directory within temp which will become the
+	// root of the archive. This replaces the `tar.ImplicitTopLevelFolder`
+	// behavior.
+	dir := filepath.Join(tmpDir, fileNameWithoutExtension(destination))
+	if err := os.Mkdir(dir, 0700); err != nil {
+		return fmt.Errorf("error creating temporary directory: %w", err)
+	}
+
+	for _, src := range files {
+		dst := filepath.Join(dir, src)
+		if err = common.CopyFile(src, dst); err != nil {
+			return fmt.Errorf("error copying file: %w", err)
+		}
+	}
+
+	tar := archiver.NewTarGz()
+	tar.OverwriteExisting = true //
+	tar.MkdirAll = true          // make destination directory if it doesn't exist
+
+	if err = tar.Archive([]string{dir}, destination); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fileNameWithoutExtension returns a filename with its extension stripped.
+func fileNameWithoutExtension(filename string) string {
+	base := filepath.Base(filename)
+	firstDot := strings.Index(base, ".")
+	if firstDot > -1 {
+		return base[:firstDot]
+	}
+	return base
+}
+
+// getIgnoredFiles reads the .fastlyignore file line-by-line and expands the
+// glob pattern into a list of files it matches. If no ignore file is present
+// it returns an empty list.
+func getIgnoredFiles(filePath string) (files []string, err error) {
+	if !common.FileExists(filePath) {
+		return files, nil
+	}
+
+	// gosec flagged this:
+	// G304 (CWE-22): Potential file inclusion via variable
+	// Disabling as we trust the source of the filepath variable as it comes
+	// from the IgnoreFilePath constant.
+	/* #nosec */
+	file, err := os.Open(filePath)
+	if err != nil {
+		return files, err
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		glob := strings.TrimSpace(scanner.Text())
+		globFiles, err := filepath.Glob(glob)
+		if err != nil {
+			return files, fmt.Errorf("parsing glob %s: %w", glob, err)
+		}
+		files = append(files, globFiles...)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return files, fmt.Errorf("reading %s file: %w", filePath, err)
+	}
+
+	return files, nil
+}
+
+// getNonIgnoredFiles walks a filepath and returns all files which aren't
+// included in the provided ignore list. The ignore list is first sorted and
+// then searched for each possible file in the tree using a binary search.
+func getNonIgnoredFiles(base string, ignoredFiles []string) ([]string, error) {
+	var files []string
+
+	if len(ignoredFiles) > 1 {
+		sort.Strings(ignoredFiles)
+	}
+
+	if err := filepath.Walk(base, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			i := sort.SearchStrings(ignoredFiles, path)
+			if found := i < len(ignoredFiles) && ignoredFiles[i] == path; !found {
+				files = append(files, path)
+			}
+		}
+		return nil
+	}); err != nil {
+		return files, err
+	}
+
+	return files, nil
+}
```

Comment:
Total nit, take it or leave it, but consider:

```go
func getNonIgnoredFiles(base string, ignoredFiles []string) ([]string, error) {
	ignore := make(map[string]bool, len(ignoredFiles))
	for _, s := range ignoredFiles {
		ignore[s] = true
	}

	err := filepath.Walk(base, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		if ignore[path] {
			return nil
		}
		files = append(files, path)
		return nil
	})
	
	return files, err
}
```

Reply:
Yeah, I considered a map though thought its construction would be slower that the sort, but it does give you `O(1)` lookups and does look cleaner. Will change :+1:

Reply:
Changed in https://github.com/fastly/cli/pull/87/commits/2e614bb6d4cf5628877e17b0a9a181ff24ee3f37